src/libfetchers/tarball.hh

```diff
@@ -24,7 +25,8 @@ DownloadFileResult downloadFile(
     const std::string & url,
     const std::string & name,
     bool locked,
-    const Headers & headers = {});
+    const Headers & headers = {},
+    bool onlyIfChanged = false);
```

Comment:
When do we want to download if not changed?

Reply:
Errr, it's fully explain in my comments above

> assuming the file at the source URL hasn't changed, when I run nix flake update, if both paths exist, all is good. nix will fetch the source URL with an etag, and get a 304 response, and do nothing. but if the source file has been deleted from the store, then nix will redownload the source file, regardless if the unpacked store path is still valid.

to summarize, if the unpacked tarball is available in store, we don't need to download the source tarball if not changed.

Reply:
err sorry I might have gotten the negation backwards. What you propose of downloading if and if needed sounds reasonable. I am asking when would one *not* want the semantics you proposed? e.g. When would one want to download again if we already have?

Reply:
personally i can't come up with a scenario where re-downloading would be preferable. do you have something in mind?

Reply:
errrr, did i misunderstand your question?

in the code, `onlyIfChanged` would be `false` for a "normal" `downloadFile` call. so say the user calls `builtins.fetchurl` then the file will be downloaded if it's unchanged but not in store.

Reply:
@yshui Sorry I am still a bit confused. It seems in both cases you are saying "if we have it in the store and it didn't change we don't need to redownload it". I am missing something --- maybe missing that `nix flake update` need not care what is in the store?

Reply:
> I am missing something

Fine. let me just list all the possible cases here.

So, there's two blobs involved for `fetchTarball`.

`tarball.tar.gz`, and `unpacked-tarball/`. for our purposes, we assume both have a entry in fetcher cache. Now, let's consider different cases where either or both of them are in the store.

1. both `tarball.tar.gz` and `unpacked-folder/` are in store.
   `fetchTarball` fetches `tarball.tar.gz` with `etag` in cache, gets 304, stops. Which is good.
2. only `tarball.tar.gz` is in store.
   `fetchTarball` fetches `tarball.tar.gz` with `etag` in cache, gets 304, unpacks the tarball already in store. This is good too.
3. neither is in store.
   `fetchTarball` ignores cache entry for `tarball.tar.gz`, redownloads it, unpacks it. Good.
4. only `unpacked-tarball/` is in store.
   a. `tarball.tar.gz` is changed at the specified url
       `fetchTarball` ignores cache entry for `tarball.tar.gz`, redownloads it, unpacks it. Good.
   b. `tarball.tar.gz` is **unchanged** at the specified url
       `fetchTarball` ignores cache entry for `tarball.tar.gz`, redownloads it, unpacks it. **NOT** good.

This PR avoids redownloading `tarball.tar.gz` in case (4.b)

OTOH, `fetchurl` only puts own blob into the store. So this is the case where `downloadFile` is called with `onlyIfChanged = false`. and the cases for `fetchurl` are:

1. file is not in store
    doesn't matter if the file at the source url is changed or not, `downloadFile` needs to download it. this is why `onlyIfChanged` should be false.
2. file is in store
    `downloadFile` downloads the file if it has changed at the source url. this is the current behavior and is unchanged.

Reply:
Thanks very much @yshui. The big thing I was forgetting what that there is both the unpack and packed versions.

@edolstra has a PR that changes how the unpacked stuff is downlodaed that we are merging soon. I suspect we should land that first.

My only question left is, why does `builtins.fetchurl` use the tarball fetcher at all? Shouldn't it just use the file fetcher?

Reply:
> why does `builtins.fetchurl` use the tarball fetcher at all?

maybe the way i phrased caused some confusion, but `builtins.fetchurl` doesn't call `downloadTarball`, it calls `downloadFile`. 

Whereas `builtins.fetchTarball` will call `downloadTarball`, and that in turn calls `downloadFile`